### DISCUSSION TITLE
Restart container from inside

### DISCUSSION THREAD

**Discussion by p0rys** [moby/moby#46682] [Q&A]
Hello everyone,

I'm currently facing a unique situation. We're binding the Docker Unix socket into a container (-v "/var/run/docker.sock:/var/run/docker.sock") and making certain API functions (container restart / start / stop) available to a WebApp through an nginx server. This allows us to start, stop, and restart all installed containers via the WebApp.

However, since updating to version 24.X.X, the restart function hasn't been working properly. When the container with the nginx server is restarted, it stops but doesn't start up again. After some investigation, I found that with this commit (https://github.com/moby/moby/commit/def549c8f6716507ee9175fe9d798a42df89e27d), the "context" of the API request is passed to the containerStart() functions. Since the request originates from the container that was just stopped, the connection closes, and the containerStart() function throws an error, indicating an incorrect context. As a result, the container stops but doesn't restart.

I understand that this is a very specific scenario, but I cannot install an nginx server on the host. So, I wanted to ask if anyone knows another way to utilize these functions (start, stop, restart of all containers) within a container. For now, as a temporary fix, we'll continue using version 23.x.x, but this doesn't seem like a sustainable solution to me.

**Comment by thaJeztah** [comment]
Thanks for reporting! I recall we had a similar issue with another endpoint (?) where cancelling the context cancelled the process; 

- https://github.com/moby/moby/issues/45731
- https://github.com/moby/moby/pull/45738

Actually wondering now if that one would also resolve the issue you're seeing; what version of docker are you running? (I see that patch was backported to the 24.0.3 patch release)

**Reply by p0rys** [reply]
Hey thaJeztah,

I came across this issue on version 24.0.2 and tested it again on 24.0.6 - the behavior remains the same. #45731 appears to match exactly with the problem we're facing, but specifically for the "stop" endpoint.

Given that we utilize the "restart" endpoint, the current fix aids only in the "first half" of the restart process. The containerStart() function still requires the context and is terminated when the context closes. I believe this error could be addressed by also incorporating a local context in the containerStart() function, similar to the merge.

**Reply by thaJeztah** [reply]
Thanks for trying! Knowing that it's not fixed with that PR is good info to have.

I think it's clear indeed what's causing the issue. I'd have to have a peek as to "what amount of cancellation is expected" but overall I think that "making the request" should be an atomic operation; once you made a successful API call to do a restart, it should complete the restart (and not cancel the action).

go1.21 has a new type of context for exactly such purposes, and I recall we implemented a version for older go versions (the 24.0 release is still using go1.20), but I'd have to check if we included that already in the 24.0 branch; https://github.com/moby/moby/pull/46582

With that (and without looking at the code yet), this _may_ be an "easy fix" if we can wrap the context in the endpoint.

**Reply by thaJeztah** [reply]
For lack of abundance (from your input here, I _think_ you have a good understanding of everything, but other readers that may arrive here may not have);

> We're binding the Docker Unix socket into a container (-v "/var/run/docker.sock:/var/run/docker.sock") and making certain API functions (container restart / start / stop) available to a WebApp through an nginx server. This allows us to start, stop, and restart all installed containers via the WebApp.

⚠️ ⚠️ **access to the docker socket is equivalent to `root` access to your host.** Mounting the socket into a container means that a compromised container will compromise your host. Exposing the API through a WebUI without additional protection can also risk exposing the API through loopback interfaces, which may make it accessible to other containers on the same host. For more information also refer to https://docs.docker.com/go/attack-surface/

**Reply by thaJeztah** [reply]
- opened a PR https://github.com/moby/moby/pull/46688
